### ISSUE TITLE
AFNetworking is not ios 6 compatible. the podspec says iOS 6 compatible

### ISSUE THREAD

**Issue by vandadschibsted** [AFNetworking/AFNetworking#2540]
NSURLSession is not available on iOS 6 at all

```objc
NS_CLASS_AVAILABLE(NSURLSESSION_AVAILABLE, 7_0)
@interface NSURLSession : NSObject
```

**Comment by mattt** [comment]
AFNetworking is indeed iOS 6 compatible. `AFURLSessionManager` & `AFHTTPSessionManager` are not available without iOS >= 7, but the rest is available.

**Comment by vandadschibsted** [comment]
Our app crashes because of the use of this constant

```objc
NSURLSessionTransferSizeUnknown
```

whose value is equal to `-1LL` and is defined in `NSURLSession.h` and is being used in `AFURLSessionManager.m` like so:

```objc
    int64_t totalUnitCount = uploadTask.countOfBytesExpectedToSend;
    if(totalUnitCount == NSURLSessionTransferSizeUnknown) {
        NSString *contentLength = [uploadTask.originalRequest valueForHTTPHeaderField:@"Content-Length"];
        if(contentLength) {
            totalUnitCount = (int64_t)[contentLength longLongValue];
        }
    }
```

If my base SDK is iOS 8 and I run my app on a device that is on iOS 6, this line will crash.

**Comment by mattt** [comment]
Like I said, you can't use that class unless you're deploying iOS >= 7.

**Comment by vandadschibsted** [comment]
are you referring to this line of code?

```objc
#if (defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && __IPHONE_OS_VERSION_MAX_ALLOWED >= 70000) || (defined(__MAC_OS_X_VERSION_MAX_ALLOWED) && __MAC_OS_X_VERSION_MAX_ALLOWED >= 1090)

@interface AFURLSessionManager : NSObject <NSURLSessionDelegate, NSURLSessionTaskDelegate, NSURLSessionDataDelegate, NSURLSessionDownloadDelegate, NSSecureCoding, NSCopying>
```

?

Because if my base SDK is 8.0 (the SDK that I build with) and my target OS is 6.0 onwards (which is perfectly acceptable), this class __will__ be available at runtime because you are checking my base SDK, not the iOS version on which the app is running.

Is there some check that you are doing which I am missing on?

**Comment by mattt** [comment]
No, I'm not referring to any line of code. I'm just restating the explicit requirements of the README.

**Comment by vandadschibsted** [comment]
It's quite unfortunately because the requirements do say that AFNetworking __is__ in fact compatible with iOS 6 but it is __not__ because it is not weakly loading the NSURLSession constants like  NSURLSessionTransferSizeUnknown therefore, upon running the code, dlyb crashes like so:

```bash
dyld: Symbol not found: _NSURLSessionTransferSizeUnknown
  Referenced from: /var/mobile/Applications/2A408740-41B8-445E-8EA2-73A7387D63A2/my.app/my
  Expected in: /System/Library/Frameworks/Foundation.framework/Foundation
 in /var/mobile/Applications/2A408740-41B8-445E-8EA2-73A7387D63A2/my.app/my
```

So that symbol is not present when running the app on iOS 6 since constants are in the section that dyld loads before even the app runs.

But if it makes you feel happier, sure, it is compatible with iOS 6 and it totally works /s

**Comment by mattt** [comment]
AFNetworking is iOS 6 compatible. You just can't use the `NSURLSession` subspec—specifically, `AFURLSessionManager` & `AFHTTPSessionManager`. If you do, you'll end up with exactly the problem you're describing. This is not unexpected.

I'm not sure how this could be made any more explicit. I'm not interested with arguing with you.

If you have a PR for how this could be made clearer in the README or elsewhere, I'd be more than happy to take a look.

**Comment by vandadschibsted** [comment]
Of course it could be made clearer :-)

it could be said that AFNetworking is __not__ compatible if you build with base SDK 7.0 or higher. Simple as that.

**Comment by gwynantj** [comment]
AFNetworking is compatible with iOS 6 whatever you set the base SDK too.....you just can't use `AFURLSessionManager` & `AFHTTPSessionManager` if your iOS Deployment Target is set to 6.

Base SDK defines the api's that are available to you during development, so if I have it set to 8.1 then I have all the API's that exists up to 8.1. But in your case you have your deployment target set to 6 but are using API's that are on available from iOS 7 onwards. So when it comes to testing on an iOS 6 device it tries to load symbols that don't exist since it's not available in the iOS 6 sdk.

So if you want to use AFNetworking and target iOS 6 don't use `AFURLSessionManager` & `AFHTTPSessionManager` or even the constants defined in `NSURLSession`.

**Comment by vandadschibsted** [comment]
Thanks @gwynantj 

What this means is that if I have an app that used iOS 8 as base SDK and supports iOS 6 onwards, I can not even import the AFURLSessionManager anywhere in my code. That means, I cannot use the "AFURLSessionManager" and the "AFHTTPSessionManager" classes on iOS 7 or 8 either.

The reason is that as soon as you import those classes in your code, the constants and other iOS 7 >= symbols will be loaded into memory and as shown before, dyld will crash because those symbols cannot be found on iOS 7.

So basically, those classes will be useless in iOS 7 and 8 if I want my app to run on iOS 6 since there is on way to conditionally, at runtime, import a class, based on the version of the OS on which your app is running.

It seems like everybody here is really trying to stick with the proposed line of "it is compatible with iOS 6" so I am happy leaving it at that and close this discussion.

Thank you both for your time.
